packages/formik/src/index.tsx

```diff
@@ -0,0 +1,202 @@
+        Object.keys(schema)
+          .filter(schemaProp => schema.hasOwnProperty(schemaProp))
+          .forEach((fieldName: string) => {
+            if (fieldName === 'id' || fieldName === 'buttons') {
+              return;
+            }
+            const schemaField = schema[fieldName];
+            if (!schemaField.type.isSchema) {
+              const fieldValue = parent ? values[parent][fieldName] : values[fieldName];
+              switch (schemaField.fieldType) {
+                case FieldTypes.input: {
+                  collector.push(
+                    this.genInput(schemaField, fieldValue, fieldName, { name: parent, value: values[parent] })
+                  );
+                  break;
+                }
+                case FieldTypes.select: {
+                  collector.push(this.genSelect(schemaField, fieldValue, fieldName));
+                  break;
+                }
+                case FieldTypes.checkbox: {
+                  collector.push(this.genCheck(schemaField, fieldValue, fieldName));
+                  break;
+                }
+                case FieldTypes.radio: {
+                  collector.push(this.genRadio(schemaField, fieldValue, fieldName));
+                  break;
+                }
+                case FieldTypes.custom: {
+                  collector.push(this.genCustomField(schemaField, fieldValue, fieldName));
+                  break;
+                }
+                default: {
+                  throw new Error(`${fieldName} has wrong field type`);
+                }
+              }
+            } else {
+              generate(schema[fieldName].type.values, fieldName, collector);
+            }
+          });
+        return collector;
+      };
+      const formElements = generate(this.schema.values, null, []);
+      formElements.push(this.genButtons(this.schema.values.buttons, isValid));
+      return (
+        <Form handleSubmit={this.handleSubmit} name={this.schema.name} input={formAttrs}>
+          {formElements}
+        </Form>
+      );
+    });
+  }
+
+  public static setValidationMessages(messages: any) {
+    DomainValidator.setValidationMessages(messages);
+  }
+
+  private getValuesFromSchema() {
+    const getValues = (schema: any, model: any) => {
+      Object.keys(schema)
+        .filter(schemaProp => schema.hasOwnProperty(schemaProp))
+        .forEach((fieldName: string) => {
+          if (fieldName === 'id') {
+            return;
+          }
+          const schemaField = schema[fieldName];
+          if (!schemaField.type.isSchema) {
+            model[fieldName] = schemaField.defaultValue || '';
+          } else {
+            model[fieldName] = {};
+            getValues(schema[fieldName].type.values, model[fieldName]);
+          }
+        });
+      return model;
+    };
+    return getValues(this.schema.values, {});
+  }
+
+  private genInput(ctx: any, value: string | number, fieldName: string, parent: any) {
```

Comment:
All of those Field components have a common structure. Why do we need 5 similar methods to generate the same component?

Reply:
Ok, I'll figure out how to improve it